src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -468,10 +469,13 @@ protected override void Dispose(bool disposing)
                 _canRead = false;
                 _canWrite = false;
                 _canSeek = false;
-                // Don't set the buffer to null, to avoid a NullReferenceException
-                // when users have a race condition in their code (ie, they call
-                // Close when calling another method on Stream like Read).
-                //_buffer = null;
+
+                if (_buffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(_buffer, clearArray: true);
```

Comment:
> I don't know enough about this area of the code to say which changes are/aren't affected

It's simple.  It's just a shared collection of ```T[]```.  You can take one out, use it, and put it back.  Some other code in your process can do the same.  If you put one back and then keep using it, someone else could take that some buffer out and be using it at the same time.